src/Features/LanguageServer/Protocol/Handler/Symbols/DocumentSymbolsHandler.cs

```diff
@@ -33,181 +30,119 @@ public DocumentSymbolsHandler()
 
         public TextDocumentIdentifier GetTextDocumentIdentifier(DocumentSymbolParams request) => request.TextDocument;
 
-        public async Task<object[]> HandleRequestAsync(DocumentSymbolParams request, RequestContext context, CancellationToken cancellationToken)
+        public async Task<SumType<DocumentSymbol, SymbolInformation>[]> HandleRequestAsync(DocumentSymbolParams request, RequestContext context, CancellationToken cancellationToken)
         {
             var document = context.Document;
             if (document == null)
             {
-                return Array.Empty<SymbolInformation>();
+                return Array.Empty<SumType<DocumentSymbol, SymbolInformation>>();
             }
 
-            var symbols = ArrayBuilder<object>.GetInstance();
+            var symbols = ArrayBuilder<SumType<DocumentSymbol?, SymbolInformation?>>.GetInstance();
 
-            var navBarService = document.Project.LanguageServices.GetRequiredService<INavigationBarItemService>();
-            var navBarItems = await navBarService.GetItemsAsync(document, cancellationToken).ConfigureAwait(false);
-            if (navBarItems.Count == 0)
+            var hierarchicalDocumentSymbolSupport = context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true;
+            var documentSymbolsService = document.Project.LanguageServices.GetRequiredService<IDocumentSymbolsService>();
+            var documentSymbols = await documentSymbolsService.GetSymbolsInDocumentAsync(
+                document,
+                hierarchicalDocumentSymbolSupport ? DocumentSymbolsOptions.FullHierarchy : DocumentSymbolsOptions.TypesAndMembersOnly,
+                cancellationToken).ConfigureAwait(false);
+            if (documentSymbols.IsEmpty)
             {
-                return symbols.ToArrayAndFree();
+                return symbols.ToArrayAndFree()!;
             }
 
             var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
             var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
 
-            // TODO - Return more than 2 levels of symbols.
-            // https://github.com/dotnet/roslyn/projects/45#card-20033869
-            if (context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true)
+            if (compilation is null || tree is null || text is null)
```

Comment:
Didn't GetSymbolsInDocumentAsync already call GetRequiredSyntaxTree and friends? So how can this happen?

Reply:
(you also already called GetRequiredService to fetch the document symbols service, so...)

Reply:
I can switch these to the Required version above.